File: javascript/sliding-window/longest-substring.js
// Longest Substring Without Repeating Characters

// Given a string s, find the length of the longest
// substring without repeating characters.

function lengthLongestSubstring(s) {
    const characters = new Set()
    let [left, count] = [0, 0]
    // iterate through every index char. in string
    for (let right = 0; right < s.length; right++) {
        // find a duplicate char.
        while (characters.has(s[right])) {
            // delete all char. from left untill we delete the duplicate char.
            characters.delete(s[left])
            // move the position of the first char. in substring
            left++
        }
        // add in substring a new char.
        characters.add(s[right])
        // compute max length
        count = Math.max(count, right - left + 1)
    }

    return count
}

// testing
console.log(lengthLongestSubstring("abcabcbb"))
// expected output: 3

File: python/sliding-window/longest_substring.py
# Longest Substring Without Repeating Characters

# Given a string s, find the length of the longest
# substring without repeating characters.

def lengthLongestSubstring(s: str) -> int:
    characters = set()
    left, length = 0, 0
    # iterate through every index char. in string
    for right in range(len(s)):
        # if we find a duplicate char. in substring
        while s[right] in characters:
            # remove char. from left untill the duplicate char. is deleted
            characters.remove(s[left])
            # move the first char. in substring to the next
            left += 1
        # add unique char. in substring
        characters.add(s[right])
        # compute the max length
        length = max(length, right - left + 1)

    return length

# testing
print(lengthLongestSubstring("abcabcbb"))
# expected output: 3
        